Re-Enable WPE_IN

FriedrichRober opened the pull request:
I think this was the issue why in version 0.7 we had a recursion depth error when testing in the gap package distro, see #67.

As suggested by @fingolfin, I will now try out a fix to re-enable the efficient membership test in wreath products.

Careful testing needs to be done, since the recursion error behaved strangely and often was hard to replicate in an interactive GAP terminal.

FriedrichRober committed:
Reactivate broken test

FriedrichRober committed:
Include ProductAction type in test for "in"

FriedrichRober committed:
Fix test: wrong spacing

FriedrichRober committed:
TryNextMethod for `in` if projection is not bound yet

FriedrichRober commented:
I think I have understood that the essential difference between the interactive GAP terminal and running the test file is the command `START_TEST`. If we remove this from the test file, it works. So to make the error appear locally instead, one needs to insert this command before running the other commands.

FriedrichRober committed:
Construct all projections and embeddings before efficient in test

FriedrichRober committed:
Add START_TEST and STOP_TEST to all human-created tests

FriedrichRober committed:
Use ComponentsOfWreathProduct instead of info.I in WPE_IN

FriedrichRober commented:
In gap lib, now also this seems to work:
```
LoadPackage("WPE");
Test("tst/testinstall/opers/ListWreathProductElement.tst");
```

FriedrichRober committed:
tst: add missing EOF

FriedrichRober commented:
I think now we should avoid any recursion error. Strictly speaking, we only need `Projection` for `WPE_IN`. I also included to generate every `Embedding` before `WPE_IN`, because they call the `WPE_IN` method and lead to a recursion.

FriedrichRober commented:
> I think I have understood that the essential difference between the interactive GAP terminal and running the test file is the command `START_TEST`. If we remove this from the test file, it works. So to make the error appear locally instead, one needs to insert this command before running the other commands.

I guess that it has something to do with Assert Levels, that `GroupHomomorphismByFunction` usually does not test for membership and with a higher assert level it does? I guess it has something to be along these lines

FriedrichRober committed:
FIX: info.constructingProjection -> info.constructingMaps

FriedrichRober commented:
@fingolfin, do you have a suggestion how I can avoid code duplication in `gap/WreathProductElements.gi` when installing the `\in` methods for permutation groups and matrix groups? Currently I do it separately, since I use `TryNextMethod()` and I think I cannot extract this outside of the scope of `InstallMethod`.

FriedrichRober committed:
DOC: put back the in-operation

fingolfin commented on gap/WreathProductElements.gi:
```diff
@@ -31,8 +31,47 @@ function(g, G)
-# InstallMethod( \in, "matrix, and matrix wreath product", true, [IsMatrix, IsMatrixGroup and HasWreathProductInfo], OVERRIDENICE + 42, WPE_IN);
+InstallMethod( \in, "perm, and perm wreath product", true,[IsPerm, IsPermGroup and HasWreathProductInfo], OVERRIDENICE + 42,
+function(g, G)
+    local info, comps, n, i;
+    info := WreathProductInfo(G);
+    comps := ComponentsOfWreathProduct(G);
+    n := NrMovedPoints(comps[2]);
+    if not IsBound(info.projection) or Length(info.embeddings) < n + 1 then
+        if not IsBound(info.constructingMaps) then
+            info.constructingMaps := true;
+            for i in [1 .. n + 1] do
+                Embedding(G, i);
+            od;
+            Projection(G);
+            info.constructingMaps := false;
+        elif info.constructingMaps then
+            TryNextMethod();
+        fi;
+    fi;
```
To avoid the code duplication, just move this code into `WPE_IN`. Yes, the `TryNextMethod()` will be fine: there is no fundamental difference between
```
InstallMethod(bla, ..., function(x) ... end);
```
and
```
foo := function(x) ... end;
InstallMethod(bla, ..., foo);
```

FriedrichRober replied:
Ah, I didn't know. Thanks!

fingolfin commented on gap/WreathProductElements.gi:
```diff
@@ -31,8 +31,47 @@ function(g, G)
+    info := WreathProductInfo(G);
+    comps := ComponentsOfWreathProduct(G);
+    n := NrMovedPoints(comps[2]);
+    if not IsBound(info.projection) or Length(info.embeddings) < n + 1 then
```
For future generations reading and debugging this code I would strongly recommend to add some comments explaining what is going on here

FriedrichRober replied:
yes, you are right. I will do it

FriedrichRober replied:
@fingolfin, I hope the comment explains the bug and fix well enough. I am thankful for any suggestions on how to improve the comment

fingolfin commented on tst/files/human-created/testNonConjugacy.tst:
```diff
@@ -11,11 +15,18 @@ gap> TestConjugacy := function(K, H, n)
 true
+
+#
+gap> STOP_TEST("testNonConjugacy.tst", 1);
```
Just FYI you can actually omit the `, 1`, it is only there for historical reasons (and most callers still have it because people copy&paste this from other packages without much thought ;-)

FriedrichRober replied:
Like I did :sweat_smile: I guess I will remove it then, no reason to stick to obsolete stuff

FriedrichRober committed:
Change STOP_TEST("...", 1) -> STOP_TEST("...")

FriedrichRober committed:
Improve WPE_IN.

- Add comment explaining the recursion bug, and the fix
- Remove code duplication

fingolfin commented on gap/WreathProductElements.gi:
```diff
@@ -20,19 +20,49 @@
+    n := NrMovedPoints(H);
+    # In order to avoid a recursion depth trap, we need to construct
+    # all embeddings and the projection before
+    # using the efficient membership in WPE_IN (after cooment [*]).
```
```suggestion
    # using the efficient membership in WPE_IN (after comment [*]).
```

FriedrichRober committed:
WPE_IN : fix typo in comment

Co-authored-by: Max Horn <max@quendi.de>

FriedrichRober committed:
Refactor dev/tests_gen code.

- Add START_TEST and STOP_TEST
- Change comments "gap> #" -> "#"
- Add more linebreaks

FriedrichRober committed:
Regenerate machine-generated tests.

- Better linebreaks
- Add START_TEST and STOP_TEST

FriedrichRober commented:
There is now an error because we use StartTest and the native command `WreathProduct(K, SymmetricGroup() )` creates a group whose generator list includes a trivial element. I will fix it now in GAP.

FriedrichRober committed:
FIX: WPE_IN, top degree was wrong for trivial top groups

FriedrichRober commented:
> There is now an error because we use StartTest and the native command `WreathProduct(K, SymmetricGroup() )` creates a group whose generator list includes a trivial element. I will fix it now in GAP.

Actually the error was caused by something else. The top group degree was wrong for trivial groups. I fixed it now.
